### PR TITLE
Fix for Rack 3 [Sequencescape]

### PR DESCRIPTION
**Pull request by StephenHulme**
Contributes towards closing #7 

Changes:
- Fix for Rack 3 compatibility in Sequencescape
- Add Ruby 3.4 and Rails 8.0 to test matrix
- Update checkout action
- Update dependencies for Rails 8.0
- Add minitest reporter
- Fix failing test setup
- Migrate error and deprecation handling to modern rails

### CODE REVIEW

File: lib/jsonapi/error.rb [hunk 1]
```diff
@@ -15,7 +15,7 @@ def initialize(options = {})
       @source         = options[:source]
       @links          = options[:links]
 
-      @status         = Rack::Utils.status_code(options[:status]).to_s
+      @status         = status_code(options[:status]).to_s
```
Comment (StephenHulme) [inline lib/jsonapi/error.rb:18]: The Rack 3 warning fix

File: lib/jsonapi/error.rb [hunk 2]
```diff
@@ -24,6 +24,17 @@ def to_hash
       instance_variables.each {|var| hash[var.to_s.delete('@')] = instance_variable_get(var) unless instance_variable_get(var).nil? }
       hash
     end
+
+    private
+
+    # Extracted from Rack 2
+    def status_code(status)
```
Comment (BenTopping) [inline lib/jsonapi/error.rb:31]: Is this needed for supporting older versions?  Because we have changed all the status codes to the correct one 'unprocessable_content' do we still need this?

File: lib/jsonapi/error.rb [hunk 3]
```diff
@@ -24,6 +24,17 @@ def to_hash
...
+      if status.is_a?(Symbol)
```
Comment (BenTopping) [inline lib/jsonapi/error.rb:36]: can status be nil?

File: lib/jsonapi/request_parser.rb [hunk 4]
```diff
@@ -272,7 +273,7 @@ def parse_sort_criteria(sort_criteria)
 
       sorts = []
       begin
-        raw = URI.unescape(sort_criteria)
+        raw = CGI.unescape(sort_criteria)
```
Comment (BenTopping) [inline lib/jsonapi/request_parser.rb:276]: Why is this needed?

**Review by BenTopping — COMMENTED** [review]
Changes look good to me.

File: lib/jsonapi/error.rb (hunk 2 above)
Comment (StephenHulme) [reply lib/jsonapi/error.rb:31]: I think this is independent of changes to how Rack works - Rack just happened to have a function that was a convenient way of translating a symbol (eg: `:bad_request`) into a status code (`400`). the problem arose since Rack made significant changes to the function, but kept the name the same.
There's an argument to be made for a more significant refactor of the error handling code, but I'm unable to look into that at the moment. Instead I have just created a snapshot of the previous implementation (see `status_code`).

File: lib/jsonapi/error.rb (hunk 3 above)
Comment (StephenHulme) [reply lib/jsonapi/error.rb:36]: I think this is possible, but it isn't being handled in the existing implementation (I've just pulled the old code out of Rack 2).

File: lib/jsonapi/request_parser.rb (hunk 4 above)
Comment (StephenHulme) [reply lib/jsonapi/request_parser.rb:276]: Good question - once I got the tests working again, it was generating a slew of warnings making the tests hard to read.
`URI.unescape` was made [obsolete in Ruby 2.5](https://ruby-doc.org/stdlib-2.5.8/libdoc/uri/rdoc/URI/Escape.html#method-i-unescape).

File: lib/jsonapi/error.rb (hunk 3 above)
Comment (BenTopping) [reply lib/jsonapi/error.rb:36]: Okay, the fallback `nil.to_i` would resolve to 0. I think its worth adding a nil guard as it looks like json api dictates what status value is used.

Comment (StephenHulme) [reply lib/jsonapi/error.rb:36]: More tests added and an error is raised when a nil status is provided: 73b5ed1213013b721fba787eab6d2fd34e44bb79

**Review by BenTopping — APPROVED** [review]

**Comment by StephenHulme** [comment]
Tested with Sequencescape in int-suite and passing.
